contracts/oracle/src/contract/alarms/iter.rs

```diff
@@ -0,0 +1,133 @@
+    }
+}
+
+/// Combines all alarms iterators, injecting price errors.
+pub struct AlarmsFlatten<'a, I> {
+    storage: &'a dyn Storage,
+    prices: I,
+    alarms: Option<AlarmsIterator<'a>>,
+}
+
+impl<'a, I> AlarmsFlatten<'a, I> {
+    pub fn new(storage: &'a dyn Storage, prices: I) -> Self {
+        Self {
+            storage,
+            prices,
+            alarms: None,
+        }
+    }
+
+    fn next_price<BaseC>(&mut self) -> Option<<Self as Iterator>::Item>
+    where
+        I: Iterator<Item = PriceResult<BaseC>> + 'a,
+        BaseC: Currency,
+    {
+        self.prices.find_map(|res_price| {
+            res_price
+                .and_then(|price| {
+                    currency::visit_any_on_ticker::<SwapGroup, _>(
+                        price.base_ticker(),
+                        AlarmsCmd {
+                            storage: self.storage,
+                            price_alarms: &MarketAlarms::PRICE_ALARMS,
+                            price: &price,
+                        },
+                    )
+                    .map(|mut alarms| {
+                        let alarm = alarms.next();
+                        self.alarms = Some(alarms);
+                        alarm.map(|res| res.map_err(ContractError::from))
+                    })
+                })
+                .unwrap_or_else(|err| Some(Err(err)))
+        })
+    }
+}
+
+impl<'a, I, BaseC> Iterator for AlarmsFlatten<'a, I>
+where
+    I: Iterator<Item = PriceResult<BaseC>> + 'a,
+    BaseC: Currency,
+{
+    type Item = Result<Addr, ContractError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.alarms
+            .as_mut()
+            .and_then(|alarms| alarms.next())
+            .map(|res| res.map_err(ContractError::from))
+            .or_else(|| self.next_price())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::super::test::test_case;
+    use super::*;
+    use crate::tests;
+    use ::currency::lease::{Atom, Cro, Weth};
+    use sdk::cosmwasm_std::{testing::MockStorage, StdError};
+
+    #[test]
+    fn error_handling() {
+        let mut storage = MockStorage::new();
+
+        test_case(&mut storage);
+
+        let mut alarms_iter = AlarmsFlatten::new(
+            &storage,
+            [
+                Ok(tests::base_price::<Weth>(1, 15)),
+                Ok(tests::base_price::<Cro>(1, 15)), // no alarms for this price
+                Ok(tests::base_price::<Atom>(1, 25)),
+                Err(StdError::generic_err("error").into()),
+            ]
+            .into_iter(),
+        );
+
+        let res = alarms_iter.next();
+        assert_eq!(Some(Ok(Addr::unchecked("recv2"))), res);
+
+        // passing empty Cro iterator
+        let res = alarms_iter.next();
+        assert_eq!(Some(Ok(Addr::unchecked("recv4"))), res);
+
+        // price error propagation
+        let res = alarms_iter.next();
+        assert_eq!(Some(Err(StdError::generic_err("error").into())), res);
```

Comment:
could you extend the test, or add another, to check if the alarms iteration continues with the next prices

Reply:
Note, in case of an error here the iteration continues. The early return is in the `try_fold` call in the `try_notify_alarms()`.